### ISSUE TITLE
Notes from the rOpenSci conference

### ISSUE THREAD

**Issue by karissa** [karissa/dat-visualdiff#1]
## Visual diffing workshop
There was a visual diffing workshop at [rOpenSci unconf](https://github.com/ropensci/unconf/issues/19), and I took some notes. Most in attendance were people who must often approach a problem where they are given tabular and they need to know how it has changed. They are mostly R users and data scientists of various types (academia, labs, industry).

--- 
did it change or not?
what changed? (constrained to a column?)
million row dataset where a diff might not be useful
summary statistics
- histogram
- medium/mean

parsing the data
i’m having to do the parsing
fix things and look at the parse data
“49,000 of 50,000” of my events changed
I’d like to see just the cells that changed
the row-baed nature of current diffing 
in a flat file
typical diffing doesn’t narrow things down enough
the notion of a column is super super important

related to data cleaning
-- take a column and do REGEX, just take out the bit that matches
-- they use open refine for this
-- there’s a refine api that can record cell by cell or row by row
— but row and column operations. theres a json standard
— maybe you could take a limited set of verbs and actions and describe that this thing happened on the column
— some limited vocabulary for data cleaning operations
— diff in a functional
— the diff is really code
— in between each stage there’s some code
— having an api is important
— e.x., set of rows dissapeared. generative description of what disappeared
— is there some limited set of things that covers 80% of things

“file salad”
— they “dump data on me”
— no one knows which one is current, which is the true
— they look the same but you’re pretty damn sure they aren’t the same
— “which one is the best”
— diff tool that looked at tabular data
— 3 four cells are edited
— we are often left with this 

comparing all of these files is difficult

in advance, track changes, vs. discovery tool
it maybe that being able to do things after the fact is more limited

collab with different people
— QAQC a dataset visually, others will automate it
— scary that its a black box
— comes back to the reproducibility
— less interested in transformations
— multiple steps of an automated data cleaning process
— diff tool like this, is a way of generating code
— there’s no code

generating the code in your head and saying ‘oh yeah, that’s what they did’

— documents everything i learned
— tabular differ it’d write the changes

closed source software world
— trifecta writes diffs as they do things
— we could use open refine to do that	

reporting
— numeric diffing
— if you want to 
— if you have the matrix of data
— if there was a transformation in a particular dimension, like a matrix composition
— detecting matrix inversion

depending on how big the dataset is, you can go through the entire table?
identical is useful because it gives you the high level summary

“there’s 4 differences but it doesn’t show you what they are”
the summary at the top
— recently manually entering a bunch of numbers
— did the data analysis that came out great
— one of the numbers came out great
— this is a ‘weird’ number in this column
— anomaly detection might be out of the scope of the diffing tool

hypothesis they’re supposed to be the same
tl;dr of what changed

pull data from the database one day, and then pull the next day, and make sure that the database didn’t change
relying on other people’s data
sometimes you expect changes but you want a succinct view of what changed
attaching this as a piece of research where a lot of times where we go to level 0 to level 4. each of those stages can be represented
— level 0 (raw data)
— level 1 (qaqc)
— level 2 (gap filling based on model)
— level 3  (insight, inference)
— level 4 (deriving some product off of it)

document comparison
— literally copy pasted everything in microsoft word just so it told me what changed
— ridiculous step
— if its very big i don’t know how that would have worked

standardizing some of the columns
— do it outside of R, you lose the functions
— open in excel and save as cvs, all of the functions got lost 
— lost track of the excel file

as we work in diverse teams we need a mechanism to verify and see what has been done. excluding teams and having technical teams do that, and data scientists do that

can do it because it works by rows
— you can imagine that you store 2 datasets one is a transformation of another
— “pick” which one seems to be more efficient
— the column based diff is more efficient
— or the other way around
— we can just transpose
— we can put it any other way for different views of data that could make it easier

once differences are isolated
— how it looks
— do you get a spreadsheet looking thing, and some are red, some are green?
— when i use a visual diffing merging tool its really nice

heatmap
— hundreds are things to see whats red, keep track and isolate them
— looking at a lot of datasets at a lot of different regions
— shading indicates activity in that column or cell
— clicking on it *could* get 
— isolate differences by cell
— at a glance you could see where are issues
UCSB

is there a point of no return?
— line by line at some point the whole file changed
— where is that line

high freq data
— check in the check that something failed
— interpreted back to the original
— outlier checks
— yam file that has different ways of checking data
— windowed, instantaneough, parses the expressions and runs them on the time serious data
— flagged returned 
— testing system for data
— I only want to use data that passes these flagged 

jenny hijacks an assertion package for that
— when i run it once, nothing there should be changing, others that change more often
— it tells her ‘wake up 

could use it to communicate who is not interested in your code to see how things changed
“when we went from raw to the clean, this is what happened”

plotting package that i’ll use to compare the values
histograms, density plots
